File: elixir/lib/fog/seam.ex
defmodule Fog.Seam do
  @moduledoc """
  Boundary between two agents' fog maps.

  High tension → agents have different blind spots → transfer potential.
  Zero tension → same fog everywhere → pure arbitrage territory.

  measure/2 is a pure function. Two maps in, one seam out.
  Pipe it:

      agent_a
      |> Fog.Map.add("orbital-mechanics", :known_unknown)
      |> then(fn fog_a -> Fog.Seam.measure(fog_a, fog_b) end)
      |> Fog.Seam.summary()
  """

  alias Fog.Map, as: FogMap

  @type t :: %__MODULE__{
    agent_a:    String.t(),
    agent_b:    String.t(),
    only_in_a:  MapSet.t(String.t()),
    only_in_b:  MapSet.t(String.t()),
    shared:     MapSet.t(String.t())
  }

  defstruct [:agent_a, :agent_b, :only_in_a, :only_in_b, :shared]

  @doc "Measure the seam between two agents' fog maps."
  def measure(%FogMap{} = map_a, %FogMap{} = map_b) do
    a_keys = map_a.gaps |> Map.keys() |> MapSet.new()
    b_keys = map_b.gaps |> Map.keys() |> MapSet.new()

    %__MODULE__{
      agent_a:   map_a.agent_id,
      agent_b:   map_b.agent_id,
      only_in_a: MapSet.difference(a_keys, b_keys),
      only_in_b: MapSet.difference(b_keys, a_keys),
      shared:    MapSet.intersection(a_keys, b_keys)
    }
  end

  @doc """
  Seam tension: fraction of gaps that are asymmetric (0.0–1.0).
  High tension → different blind spots → transfer potential.
  """
  def tension(%__MODULE__{} = seam) do
    total = MapSet.size(seam.only_in_a) + MapSet.size(seam.only_in_b) + MapSet.size(seam.shared)
    case total do
      0 -> 0.0
      _ -> (MapSet.size(seam.only_in_a) + MapSet.size(seam.only_in_b)) / total
    end
  end

  @doc "Gaps neither agent can fill — need external signal."
  def system_gaps(%__MODULE__{} = seam), do: seam.shared

  @doc "Interpretation based on tension score."
  def interpretation(%__MODULE__{} = seam) do
    case tension(seam) do
      t when t > 0.7 -> "high-potential seam — asymmetric blind spots, strong transfer signal"
      t when t > 0.4 -> "active seam — partial overlap, exchange probable"
      t when t > 0.0 -> "low-tension seam — mostly shared fog, limited exchange value"
      _              -> "flat seam — identical fog, no transfer possible"
    end
  end

  @doc "Human-readable summary."
  def summary(%__MODULE__{} = seam) do
    t = Float.round(tension(seam), 2)
    "FogSeam(#{seam.agent_a}↔#{seam.agent_b}) " <>
    "tension=#{t} " <>
    "A-only=#{MapSet.size(seam.only_in_a)} " <>
    "B-only=#{MapSet.size(seam.only_in_b)} " <>
    "shared=#{MapSet.size(seam.shared)} " <>
    "— #{interpretation(seam)}"
  end
end

File: haskell/src/Fog/Seam.hs
-- | FogSeam — boundary between two agents' fog maps.
--
-- measure is a pure function. Two maps in, one seam out.
-- tension tells you whether the ignorance is asymmetric (transfer potential)
-- or symmetric (pure arbitrage territory).
module Fog.Seam
  ( FogSeam(..)
  , SeamInterpretation(..)
  , measure
  , tension
  , systemGaps
  , interpret
  , summarise
  ) where

import Data.Set (Set)
import qualified Data.Set as Set
import Fog.Map (FogMap, fogKeys, fogAgentId)

-- | What the seam is telling you.
data SeamInterpretation
  = HighPotential  -- ^ Asymmetric blind spots — strong transfer signal.
  | ActiveSeam     -- ^ Partial overlap — exchange probable.
  | LowTension     -- ^ Mostly shared fog — limited exchange value.
  | FlatSeam       -- ^ Identical fog — no transfer possible.
  deriving (Eq, Show)

-- | Boundary between two agents' fog states.
data FogSeam = FogSeam
  { seamAgentA  :: String
  , seamAgentB  :: String
  , onlyInA     :: Set String  -- ^ A dark, B may have signal.
  , onlyInB     :: Set String  -- ^ B dark, A may have signal.
  , seamShared  :: Set String  -- ^ Both dark — needs external signal.
  } deriving (Eq, Show)

-- | Measure the seam between two agents' fog maps. Pure function.
measure :: FogMap -> FogMap -> FogSeam
measure mapA mapB = FogSeam
  { seamAgentA = fogAgentId mapA
  , seamAgentB = fogAgentId mapB
  , onlyInA    = Set.difference ak bk
  , onlyInB    = Set.difference bk ak
  , seamShared = Set.intersection ak bk
  }
  where
    ak = fogKeys mapA
    bk = fogKeys mapB

-- | Seam tension: fraction of gaps that are asymmetric.
-- High tension → different blind spots → transfer potential.
tension :: FogSeam -> Double
tension s
  | total == 0 = 0.0
  | otherwise  = fromIntegral asymmetric / fromIntegral total
  where
    asymmetric = Set.size (onlyInA s) + Set.size (onlyInB s)
    total      = asymmetric + Set.size (seamShared s)

-- | Gaps that need external signal — neither agent can fill from the other.
systemGaps :: FogSeam -> Set String
systemGaps = seamShared

-- | Classify the seam.
interpret :: FogSeam -> SeamInterpretation
interpret s
  | t > 0.7   = HighPotential
  | t > 0.4   = ActiveSeam
  | t > 0.0   = LowTension
  | otherwise = FlatSeam
  where t = tension s

-- | Human-readable interpretation.
interpretLabel :: SeamInterpretation -> String
interpretLabel HighPotential = "high-potential seam — asymmetric blind spots, strong transfer signal"
interpretLabel ActiveSeam    = "active seam — partial overlap, exchange probable"
interpretLabel LowTension    = "low-tension seam — mostly shared fog, limited exchange value"
interpretLabel FlatSeam      = "flat seam — identical fog, no transfer possible"

-- | Human-readable summary.
summarise :: FogSeam -> String
summarise s =
  "FogSeam(" <> seamAgentA s <> "↔" <> seamAgentB s <> ") "
  <> "tension=" <> show (round2 (tension s)) <> " "
  <> "A-only=" <> show (Set.size (onlyInA s)) <> " "
  <> "B-only=" <> show (Set.size (onlyInB s)) <> " "
  <> "shared=" <> show (Set.size (seamShared s)) <> " "
  <> "— " <> interpretLabel (interpret s)
  where
    round2 x = fromIntegral (round (x * 100) :: Int) / 100.0 :: Double
